### DISCUSSION TITLE
Help sort out Ivo rotations

### DISCUSSION THREAD

**Discussion by Markemp** [Markemp/Cryengine-Converter#96] [Q&A] [answered]
Trying to sort out the rotations/translations (aka transform matrix) for the Ivo files.  I have a parent bone (World) and child bone (hip).  The parent bone is at `0, 0, 0` with a rotation quaternion as `0, 0, 0, 1`.  Basically your bone at the origin with no rotation.

![image](https://user-images.githubusercontent.com/12057128/128114296-753a9f77-cd21-4aa8-a2e5-7174c734bd56.png)

The hip bone is the child to that.  In the world, it should be at `0, 0, 1.00007`.  Basically straight up from the root bone (z direction is up).  It has both the relative and world quaternion as `0, -0.707, 0, 0.707`.

![image](https://user-images.githubusercontent.com/12057128/128114455-9a174ab7-08df-44a5-a6ef-345f16fd4d33.png)

There are 2 different transform (4x4) matrices I need to worry about:  The Bind Pose Matrix and the node matrix.  For all other models, the Bind Pose Matrix consists of the WorldToBone matrix in the models (a 3x4 matrix I can convert to 4x4 by just adding 0, 0, 0, 1) with no additional math.  Seems to match up with the models properly.

The BoneToWorld matrix gets me to the proper bone positions with some relatively straightforward math:

![image](https://user-images.githubusercontent.com/12057128/128114894-5d99e8e7-531e-4e44-a729-f31638891827.png)

It seems like both the BPM and the node's transform matrix have an affect on the bone going to the right place in the Collada file, but I'm not really sure how.

The question is:  For #ivo files, what is the correct BPM and node transform matrix, and how do I get there?

This affects the armatures for all #ivo files, and really needs to be sorted out before I feel comfortable releasing 1.3.1. :(

**Comment by Kjasi** [comment]
This little tid-bit may help you... .chr files import their bones in the correct positions and rotations.

To test this, I compared a chr file to it's associated skin file, specifically `Data\Objects\Spaceships\Thrusters\DRAK\DRAK_Buccaneer_Main\DRAK_Buccaneer_Main_Nozzle_Skel.chr` and `DRAK_Buccaneer_Main_Nozzle_Skin.skin` files.

![image](https://user-images.githubusercontent.com/5256123/128178975-9ca4ec79-0c00-455e-be05-69794d46969c.png)
On the left, the .chr file, and on the right, the skin file. Both are from 3.14, both are ivo files, but only the chr file has the correct positions for the bones.

Just to make sure positions AND rotations were correct, I opened a humanoid chr file (`Data\Objects\Characters\Human\tagaca\tagaca_skl.chr` and was pleasantly surprised when I had auto connect and link bones enable:

![image](https://user-images.githubusercontent.com/5256123/128181088-2492335c-814d-4695-80b3-8b560c21e111.png)
As you can see, it's in a humanoid shape, and the bones DO have rotation, and even the hand's rotations seem correct.

This means that you already have the transform code needed to export the correct bone placement and rotation. So, the question is, what's different about the chr file's bones from the skin file's bones?

Hope this helps!

**Reply by Markemp** [reply]
That actually helps a ton, thank you!  The thruster is a great model for me to investigate as it's not overly complicated.  Now to figure out why they are so different.

**Reply by Markemp** [reply]
Comparing the compiled bones between the two models, and the rotation quaternion and translations are _exactly the same_!  Curious.  Why does the exporter work for the skeleton but not the skin? 🤔

**Reply by Kjasi** [reply]
Well, my first guess is that something in Controller-Bind-Pose is wrong.

Diffing the exported dae file shows that the chr and the skin files are the same, except for library_controllers node and the library_geometries existing in the skin file. If we safely ignore geos, and everything not tied to (what I'm assuming is) bone positions, that brings us to Controller-Bind-Pose.

So how does that data compare to the real data?

**Reply by Markemp** [reply]
Good question. I'm going to have to do some testing tonight.  I wonder if I just replace the BPM (Controller-Bind-Pose) with identity matrices if that will solve the problem.  The BPM seems correct for all other models, so... maybe it's wrong for all other models, but wrong in a way that looks right in Blender?

In any event, this was a big step forward, so will keep testing.  Thanks again!

**Reply by lonelyhawk** [reply]
@Markemp Few questions here: 
1) why are you so actively refusing the matrix-based operations?
2) ignore the scale part of it? 
Take your example:
![image](https://user-images.githubusercontent.com/12057128/128114894-5d99e8e7-531e-4e44-a729-f31638891827.png)
This is a few samples of what I use in my local branch instead and it works just as fine.
![image](https://user-images.githubusercontent.com/16024236/134603596-8ecaf8a5-3c18-4143-b08f-3ddeebdf2580.png)
![image](https://user-images.githubusercontent.com/16024236/134603970-87eb614a-ac19-474d-9d55-a16f063c535e.png)

**Reply by Markemp** [reply]
It's usually a result of not knowing what the heck I'm doing, so I iterate off something that may work, test it, and... then fail the next step to find a better way to do it. :)

The scale stuff I generally ignore as I don't think I've ever seen a model that has scaling info in the model definition.  I think that games usually have a default model, and if they need a bigger version (more powerful grunt, mini boss based off the same model), then that is where the scaling is applied to the default matrix.  I also seem to get the scale and translation parts of the matrix confused... like, is scale the bottom row or last column?  There is a Matrix3x4 structure in Cryengine that ignores scale, so the last column has to be the translation, but... isn't it supposed to be the bottom row? I just can't keep them straight, and it drives me nuts.

A little history on the converter:  I was just trying to find a way to make giant dancing robots for my videos from MechWarrior Online.  Most of the early math was just trial and error on how to handle inherited transform matrices, since I didn't have any exposure on how that stuff worked.  And as in the collada code you linked, I was pretty new to coding, so I missed out on a lot of tools that could have simplified my life.  You can probably ask @peter-dolkens about what my early code looked like for a fun (although traumatic) trip down memory lane. :)

I would *love* some help with the matrix calculations in the project.  I have a lot on my plate (interview with Amazon tomorrow morning! 🤞🏽 ), so if you have some optimizations and fixes, send in the pull requests.  The current logic (mostly) works for the wrong reasons, but I really want it to always work for the right reasons, and need help with that.

**Reply by lonelyhawk** [reply]
it does get crazy with Crytek. E.g. the local transform matrix for the meshes is 3x4 stored as 4x4, with the extra fields... ignored!

**Reply by Markemp** [reply]
Like... I get in principle why you don't want to store data you don't need.  But that seems like a huge case of premature optimization.  The number of transform matrices in a typical file is tiny relative to the size of the file.  But now they need to support a 3x4 matrix along with all the operations it will need.  WHY???

... Don't even get me started on their non-standard implementation of the Half float data structure. 😡

**Reply by lonelyhawk** [reply]
Just have a look at https://github.com/CRYTEK/CRYENGINE/blob/release/Code/CryEngine/Cry3DEngine/CGF/CGFLoader.cpp from time to time.
Also, in regards of the CryHalf float data type. you treat it as 8 bit mantissa +8  bits exponent. But   https://github.com/CRYTEK/CRYENGINE/blob/release/Code/CryEngine/CryCommon/CryMath/CryHalf.inl shows the use of 10+6. Is this smth specific to SC or what?

**Reply by Markemp** [reply]
Oh crap, that's right.  It was SC/CIG that created their own Half structure.  I think the CryEngine uses the standard IEEE 754 Half data structure (1 sign bit, 5 exponent bits, and 10 significand bits.  SC has another implementation (8 + 8) that isn't standard, and it was Dymek who sussed that out originally.  ... Maybe the CryHalf should be a CIGHalf instead?  I think that's a CIG thing.

The source code has been invaluable.  A lot of the cleanup I have done is based off of the code from CryTek and Amazon (Lumberyard).  It has clarified a lot of the questions I have, but c++ code is still challenging to parse.  Visual Studio does work well with it at least, which helps.

Is the 8+8 Half structure more performant than the 6+10 version?  Provide additional needed precision?  Who knows?  Creating proprietary data structures can solve some problems, but supporting them can be a bit of a nightmare (looking at you, #ivo files).

**Reply by Kjasi** [reply]
8+8... That solves a question I had about the file formats for CIG morph targets... I'll have to deal with that later...

**Reply by lonelyhawk** [reply]
Are you talking about the Byte2HexIntFracToFloat2 function? It does something totally crazy. The first byte is an integer part, and the second part is a fraction. So 0x1234 => 0x12 (18 dec) + 0x34 (0.203125 dec)
I don't know who's more crazy here: Dymek, for figuring that out, or CIG for the invention. No way it is any performant.

**Comment by Markemp** [comment, marked as answer]
Ok, figured out a TON of stuff over the past few weeks.

1)  CryHalf vs IEEE Half vs Dymek Half

    // CryHalf Max value: 131008, min value: -130944
    // Dymek Half max value: 1.0078433, min value: -1.007874
    // Half max value: 65500, min value: -65500

😒

The Dymek half kinda makes sense with the scaling hack that SC is doing.  The scale is set to the bounding box dimensions, as all the vertices are (sorta) normalized between -1 and 1.

CryHalfs are just weird, and although they exist in the code I'm not sure if I've seen any models that use them. 

There are regular IEEE Halfs used in lots of places (non SC).  I think most studios just uses floats like normal devs.  

2)  Ivo file rotations

Finally figured these out.  There are 2 quaternions with vectors in the bone files.  These contain the local transform and the world transform of the bones.  In more traditional Cryengine files, they are stored as 3x4 matrices (`LocalTransformMatrix` and `WorldTransformMatrix`).  The `LocalTransformMatrix` contained the actual bind pose matrix, which you can just slap into the Collada binding for the controller.  The other contains the location/rot of the bone in world space, and I'm not using it for anything.

For the Ivo files, turns out the 2nd quat/vector pair contains the world transform info.  So first I make a 4x4 matrix from that quaternion, set the translation in that matrix, and then invert that matrix.  That gives the same values as the `LocalTransformMatrix`.

How did I finally figure that out?  I asked ChatGPT. 🤣 

```
Given a quaternion and a vector in world space, how do I get the local space for that point?

To get the local space for a given quaternion and vector in world space, you must first convert the quaternion to a 4x4 matrix. Then, you can use the inverse of the matrix to transform the vector from world space to local space.
```

Anyway, I should have a new version coming out soon (`v1.5.0`).

**Reply by Kjasi** [reply]
I've been using ChatGPT to help me with my programming too...

**Comment by Markemp** [comment]
Oh man... oh man... 🤤
![image](https://user-images.githubusercontent.com/12057128/210281947-8737aa9a-c916-4e9f-b876-98d6273e63de.png)
